### PR TITLE
ruby: 2.5.3 -> 2.6.1

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

I tested this with nix-review a little while ago and everything built okay, but breakages are likely to occur at runtime anyway.

Since I'm not going to go through every Ruby package one at a time, I think the best thing to do is bump this so we can get bug reports of breakages from people who actually use those packages.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by thefloweringash** [comment]
FYI, Ruby 2.6.0 has a known bug that affects http. Details at [Ruby Bug #15468](https://bugs.ruby-lang.org/issues/15468) and https://github.com/ruby/ruby/pull/2058.

**Review by zimbatm — APPROVED** [review]

**Comment by alyssais** [comment]
https://github.com/rubygems/rubygems/issues/2587 affects us also – (some?) gems don't seem to build using our `buildRubyGem` stuff because of a regression in RubyGems. It's been fixed in master, but we probably want to wait for a Ruby release that includes a RubyGems release with the fix.

**Comment by alyssais** [comment]
Also https://github.com/lsegal/yard/issues/1218 seems to affect quite a few gems if they're built with documentation, but that's easy to work around.

**Comment by alyssais** [comment]
I just merged #54939, but it doesn’t include the RubyGems fix. I don’t think we should merge this without it. Perhaps we could just update the RubyGems we ship with Ruby…

**Comment by lheckemann** [comment]
@GrahamcOfBorg test redmine gitlab

These use ruby, I guess?

**Comment by Mic92** [comment]
Gitlab is best to be tested non on ofborg. The test timeouts because it loads a lot of of files via qemu's p9 files systems.

**Comment by lheckemann** [comment]
No, in this instance it's a real failure in dependencies.

**Comment by alyssais** [comment]
Aforementioned RubyGems upgrade is #60394. After that's merged, this is
good to go from my PoV.

**Comment by alyssais** [comment]
I’ve been using this in my config for ages now, and haven’t seen any problems that were not addressed by #60394, so I intend to merge this in the next couple of days unless anyone objects.

**Comment by petabyteboy** [comment]
I tried using this but saw the problem with redis-rack from #60676 again. I manually cherry-picked the rubygems update from staging and this.
If it doesn't affect any existing packages this shouldn't prevent you from merging this, but I suspect the problem with https://github.com/rubygems/rubygems/issues/2587 is still present.

**Comment by alyssais** [comment]
Thanks @petabyteboy. Given how annoying that issue is to work around for the people who are most likely to benefit from this change, I’ll hold off merging this for now and see what can be done about that issue.

**Comment by alyssais** [comment]
So, I definitely think we should hold off on this until the gem build
issue is addressed.

To address that, we have a few options:

1. Apply that patch to our packaged RubyGems 3.0.3. This is what Fedora
   has done.
2. Wait for a new release of RubyGems that includes the fix.
3. Wait for a new release of RubyGems that includes the fix, and then a
   new release of Ruby that includes that version of RubyGems.

On one hand, it's probably not hurting anybody that Ruby 2.5 is the
default instead of 2.6. On the other, we should probably fix Ruby 2.6
anyway for people who need that version, and if we're going to do that
(by backporting the patch), we'll have a working 2.6 so might as well
update the default anyway.

After thinking about this, I think we should backport the fix to
RubyGems 3.0.3, and then make Ruby 2.6 the default Ruby.

I'd like to hear opinions from at least one other person before I do
anything though, so: thoughts?

**Comment by petabyteboy** [comment]
> After thinking about this, I think we should backport the fix to
RubyGems 3.0.3, and then make Ruby 2.6 the default Ruby.

Agreed.

**Comment by petabyteboy** [comment]
Unfortunately the problem was still present, even with the patch applied:
https://github.com/petabyteboy/nixpkgs/commits/feature/ruby
Is there anything obvious I'm missing?

**Comment by alyssais** [comment]
> Unfortunately the problem was still present, even with the patch applied:
> https://github.com/petabyteboy/nixpkgs/commits/feature/ruby

@petabyteboy would you mind applying #68684 to your tree and letting me
know if that resolves the problem?

@manveru same goes for you and the 2.6 issue you've been seeing.

**Comment by petabyteboy** [comment]
With all commits in staging concerning ruby and those from the PR you mentioned cherry-picked I still run into the same problem.

**Comment by alyssais** [comment]
> Edit: Nevermind, with this tree it worked: [petabyteboy/nixpkgs:ruby@`feature` (commits)](https://github.com/petabyteboy/nixpkgs/commits/feature/ruby)

Would you mind posting this sort of thing as a separate comment in future? I didn't see this until now because GitHub doesn't notify for edits, and had spent days wondering how it could still be broken.

**Comment by petabyteboy** [comment]
Sorry, I will do that if I remember. There are no issues anymore from my side, as long as we apply the patch for rubygems#2587 before updating.

**Comment by petabyteboy** [comment]
Is there any new or existing reason not to merge this?

**Comment by manveru** [comment]
I think before we merge this, we should upgrade ruby to 2.6.3 as well as rubygems and see if that addresses the issue.

**Comment by petabyteboy** [comment]
ruby_2_6 is already on 2.6.3 in unstable. 
The rubygems version 3.0.3 should already have the fix, but 3974254f6a20191b6b07621f6ea928be75dad121 and dfe7e42ff5844a4fe9901ce1be6125127085e044 are required to make use of our own rubygems. The rubygems included in ruby 2.6.3 still has the issue.
I could build mastodon successfully with https://github.com/petabyteboy/nixpkgs/commits/feature/ruby-testing.

**Comment by lheckemann** [comment]
Friendly reminder that the branch-off is in 11 days — do you want this in 19.09 and what do you think the chances of it being ready on time are?

**Comment by petabyteboy** [comment]
I think it's very realistic to make it before the branch-off. All the fixes have been committed since months. I think if we can merge it before that would be great, but if it doesn't work out that's not too bad either.

**Comment by alyssais** [comment]
Closing in favour of #67597.

**Pull request closed without merging**
